### ISSUE TITLE
deriving(Eq) requires Eq for type parameter even when phantom

### ISSUE THREAD

**Issue by kmcallister** [rust-lang/rust#7671]
Code:

~~~.rs
#[deriving(Eq)]
struct Foo<T> {
    x: int,
}

struct Phantom;

fn main() {
    let x : Foo<Phantom> = Foo { x: 0 };
    println((x == x).to_str());
}
~~~

Result:

~~~
$ rust -v
rust 0.7 (a2db7c1 2013-07-02 09:25:44 -0700)
host: x86_64-unknown-linux-gnu
$ rust run foo.rs 
foo.rs:10:13: 10:19 error: failed to find an implementation of trait std::cmp::Eq for Phantom
foo.rs:10     println((x == x).to_str());
~~~

**Comment by cristicbz** [comment]
This hit me today with:
```rust
#[derive(Debug)] struct Foo;
struct Bar;

#[derive(Debug)] struct Baz<T>(u64);

fn main() {
    println!("{:?}", Baz::<Foo>(10));
    println!("{:?}", Baz::<Bar>(10));  // Does not compile.
}
```

**Comment by Kimundi** [comment]
Now that Rust has `where` clauses, this could probably be resolved by emitting a `where` clause for each field of a struct that contains a type parameter:

```rust
struct NoClone;
struct Foo<A, B, C> {
    a: u8,
    b: Option<B>,
    c: C
}

// CURRENT, wrong way to implement clone in deriving
impl<A: Clone, B: Clone, C: Clone> Clone for Foo<A, B, C> {
    fn clone(&self) -> Self {
        Foo {
            a: Clone::clone(&self.a),
            b: Clone::clone(&self.b),
            c: Clone::clone(&self.c),
        }
    }
}

// NEW, Right way to implement clone in deriving
impl<A, B, C> Clone for Foo<A, B, C> where Option<B>: Clone,
                                           C: Clone {
    fn clone(&self) -> Self {
        Foo {
            a: Clone::clone(&self.a),
            b: Clone::clone(&self.b),
            c: Clone::clone(&self.c),
        }
    }
}
```

**Comment by goffrie** [comment]
The above is implemented in #21237, but unfortunately it seems to have a problem: what if a type inside the struct is a private? You can't put a bound on a private type on the trait impl, because the user of your crate (say) doesn't know what impls exist for the private type :(
